src/NHibernate/Engine/Query/HQLQueryPlan.cs

```diff
@@ -93,60 +94,52 @@ public void PerformList(QueryParameters queryParameters, ISessionImplementor ses
-				selection.Timeout = queryParameters.RowSelection.Timeout;
-				queryParametersToUse = queryParameters.CreateCopyUsing(selection);
+				if (queryParameters.RowSelection.FirstRow != RowSelection.NoValue)
+					items = items.Skip(queryParameters.RowSelection.FirstRow);
+
+				if (queryParameters.RowSelection.MaxRows != RowSelection.NoValue)
+					items = items.Take(queryParameters.RowSelection.MaxRows);
 			}
-			else
+
+			ArrayHelper.AddAll(combinedResults, items);
+		}
+
+		private QueryParameters GetModifiedQueryParametersForInMemoryPaging(QueryParameters originalParameters, int alreadyTraversedRecords)
+		{
+			RowSelection selection = new RowSelection();
+			selection.FetchSize = originalParameters.RowSelection.FetchSize;
+			selection.Timeout = originalParameters.RowSelection.Timeout;
+			if (originalParameters.RowSelection.MaxRows != RowSelection.NoValue)
 			{
-				queryParametersToUse = queryParameters;
+				selection.MaxRows = Math.Max(0, (originalParameters.RowSelection.FirstRow == RowSelection.NoValue ? 0 : originalParameters.RowSelection.FirstRow) 
+												+ originalParameters.RowSelection.MaxRows
+												- alreadyTraversedRecords);
 			}
+			return originalParameters.CreateCopyUsing(selection);
+		}
 
-			IList combinedResults = results ?? new List<object>();
-			IdentitySet distinction = new IdentitySet();
-			int includedCount = -1;
-			for (int i = 0; i < Translators.Length; i++)
+		private IEnumerable<IList> GetResultSets(QueryParameters queryParameters, ISessionImplementor session, bool needsInMemoryPaging)
+		{
+			Func<int, QueryParameters> getQueryParametersToUse = alreadyTraversedRecords => queryParameters;
+			if (needsInMemoryPaging)
 			{
-				IList tmp = Translators[i].List(session, queryParametersToUse);
-				if (needsLimit)
-				{
-					// NOTE : firstRow is zero-based
-					int first = queryParameters.RowSelection.FirstRow == RowSelection.NoValue
-												? 0
-												: queryParameters.RowSelection.FirstRow;
-
-					int max = queryParameters.RowSelection.MaxRows == RowSelection.NoValue
-											? RowSelection.NoValue
-											: queryParameters.RowSelection.MaxRows;
+				getQueryParametersToUse = alreadyTraversedRecords => GetModifiedQueryParametersForInMemoryPaging(queryParameters, alreadyTraversedRecords);
+			}
 
-					int size = tmp.Count;
-					for (int x = 0; x < size; x++)
-					{
-						object result = tmp[x];
-						if (distinction.Add(result))
-						{
-							continue;
-						}
-						includedCount++;
-						if (includedCount < first)
-						{
-							continue;
-						}
-						combinedResults.Add(result);
-						if (max >= 0 && includedCount > max)
-						{
-							// break the outer loop !!!
-							return;
-						}
-					}
-				}
-				else
-					ArrayHelper.AddAll(combinedResults, tmp);
+			int traversedRecords = 0;
+			for (int i = 0; i < Translators.Length; i++)
+			{
+				IList tmp = Translators[i].List(session, getQueryParametersToUse(traversedRecords));
```

Comment:
Because of the line 122 above, this may execute translators with a `MaxRows` of query parameter equal to 0, which is useless. Could this condition be checked for issuing a `yield break;` instead?